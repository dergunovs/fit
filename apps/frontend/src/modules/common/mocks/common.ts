import { vi } from 'vitest';

export const spyCopyToClipboard = vi.spyOn(navigator.clipboard, 'writeText').mockResolvedValue();
