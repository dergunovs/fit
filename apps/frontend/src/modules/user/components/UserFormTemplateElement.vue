<template>
  <UiChip>
    <span data-test="user-form-template-title">
      {{ props.template.title }}
    </span>

    <button @click="emit('edit', template)" type="button" :class="$style.edit" data-test="user-form-template-edit">
      <IconEdit width="20" height="20" />
    </button>

    <UiClose
      v-if="props.template._id"
      @click="emit('delete', props.template._id)"
      isSmall
      isDelete
      data-test="user-form-template-delete"
    />
  </UiChip>
</template>

<script setup lang="ts">
import { UiChip, UiClose } from 'mhz-ui';
import { IUserTemplate } from 'fitness-tracker-contracts';

import IconEdit from '@/common/icons/edit.svg';

interface IProps {
  template: IUserTemplate;
}

interface IEmit {
  edit: [template: IUserTemplate];
  delete: [id: string];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();
</script>

<style module lang="scss">
.edit {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 2px;
  color: var(--color-black);
  cursor: pointer;
  -webkit-user-select: none;
  user-select: none;
  background: none;
  border: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-touch-callout: none;
}
</style>
