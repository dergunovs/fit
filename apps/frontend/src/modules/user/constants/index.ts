export const URL_USER = '/users';
export const URL_USER_CREATE = `${URL_USER}/create`;
export const URL_USER_EDIT = `${URL_USER}/edit`;
export const URL_USER_PROFILE = `${URL_USER}/profile`;

export const USER_GOALS_OPTIONS = {
  activities: [1, 2, 3, 4, 5],
  sets: [12, 16, 20, 24, 30],
  repeats: [8, 10, 12, 15, 20],
  duration: [20, 30, 40, 50, 60],
};
