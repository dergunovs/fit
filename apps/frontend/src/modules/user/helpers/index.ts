import { IEquipment, IUserEquipment } from 'fitness-tracker-contracts';

export function excludeChoosenUserEquipment(equipments: IEquipment[], userEquipments?: IUserEquipment[]) {
  return equipments.filter(
    (equipment) => !userEquipments?.some((equipmentToFilter) => equipment._id === equipmentToFilter.equipment?._id)
  );
}
