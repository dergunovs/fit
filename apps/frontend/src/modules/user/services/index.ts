import { ComputedRef, Ref } from 'vue';
import {
  API_USER,
  API_USER_PASSWORD,
  TGetUsersDTO,
  TGetUsersQueryDTO,
  TGetUserDTO,
  TPostUserDTO,
  TPostUserDataDTO,
  TUpdateUserDTO,
  TUpdateUserDataDTO,
  TDeleteUserDTO,
  TUpdateUserPasswordDTO,
  TUpdateUserPasswordDataDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const userService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_USER, page],
      queryFn: async () => {
        const params: TGetUsersQueryDTO = { page: page.value };

        const { data } = await api.get<TGetUsersDTO>(API_USER, { params });

        return data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined> | Ref<string | undefined>) => {
    return useQuery({
      queryKey: [API_USER, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetUserDTO>(`${API_USER}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (formData: TPostUserDataDTO) => {
        const { data } = await api.post<TPostUserDTO>(API_USER, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (formData: TUpdateUserDataDTO) => {
        const { data } = await api.patch<TUpdateUserDTO>(`${API_USER}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  updatePassword: (options: object, id?: string) => {
    return useMutation({
      mutationKey: [API_USER_PASSWORD, id],
      mutationFn: async (password: TUpdateUserPasswordDataDTO) => {
        if (!id) return null;

        const { data } = await api.patch<TUpdateUserPasswordDTO>(`${API_USER_PASSWORD}/${id}`, password);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteUserDTO>(`${API_USER}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
