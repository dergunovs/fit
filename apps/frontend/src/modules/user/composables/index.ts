import { computed, Ref } from 'vue';

import { useTI18n } from '@/common/composables';

export function useUserFormTabs(isEdit: boolean, isAdmin: Ref<boolean>) {
  const { t } = useTI18n();

  const generalTab = computed(() => {
    return { value: 'general', title: t('user.general') };
  });

  const exercisesTab = computed(() => {
    return { value: 'exercises', title: t('exercise.many') };
  });

  const otherTabs = computed(() => [
    { value: 'goals', title: t('goals') },
    { value: 'weight', title: t('weight') },
    { value: 'templates', title: t('template.many') },
  ]);

  const TABS = computed(() => {
    if (!isEdit) return [generalTab.value];
    if (isAdmin.value) return [generalTab.value, ...otherTabs.value];

    return [generalTab.value, ...otherTabs.value, exercisesTab.value];
  });

  return { TABS };
}
