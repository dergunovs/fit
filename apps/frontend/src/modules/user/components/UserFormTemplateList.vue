<template>
  <UiFlex wrap>
    <UserFormTemplateElement
      v-for="template in props.templates"
      :key="template._id"
      :template="template"
      @edit="(template) => emit('edit', template)"
      @delete="(id) => emit('delete', id)"
      data-test="user-form-template-element"
    />
  </UiFlex>
</template>

<script setup lang="ts">
import { IUserTemplate } from 'fitness-tracker-contracts';
import { UiFlex } from 'mhz-ui';

import UserFormTemplateElement from '@/user/components/UserFormTemplateElement.vue';

interface IProps {
  templates: IUserTemplate[];
}

interface IEmit {
  edit: [template: IUserTemplate];
  delete: [id: string];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();
</script>
