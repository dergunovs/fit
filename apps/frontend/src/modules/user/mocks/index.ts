import { vi } from 'vitest';
import {
  TDeleteUserDTO,
  TGetUsersDTO,
  TPostUserDataDTO,
  TPostUserDTO,
  TUpdateUserDataDTO,
  TUpdateUserDTO,
} from 'fitness-tracker-contracts';

import { userService } from '@/user/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { USER_FIXTURE, USERS_FIXTURE } from '@/user/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetUser = vi.spyOn(userService, 'getOne').mockReturnValue(mockQueryReply(USER_FIXTURE));

const mockGetUsersData: TGetUsersDTO = { data: USERS_FIXTURE, total: USERS_FIXTURE.length };

const spyGetUsers = vi.spyOn(userService, 'getMany').mockImplementation(() => mockQueryReply(mockGetUsersData));

const spyCreateUser = vi.fn();
const spyUpdateUser = vi.fn();
const spyDeleteUser = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(userService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostUserDTO, TPostUserDataDTO>(spyCreateUser);
});

vi.spyOn(userService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateUserDTO, TUpdateUserDataDTO>(spyUpdateUser);
});

vi.spyOn(userService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteUserDTO, string>(spyDeleteUser);
});

export { spyGetUser, spyGetUsers, spyCreateUser, spyUpdateUser, spyDeleteUser, mockOnSuccess, mockGetUsersData };
