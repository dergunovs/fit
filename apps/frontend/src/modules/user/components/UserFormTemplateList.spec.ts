import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import UserFormTemplateList from './UserFormTemplateList.vue';
import UserFormTemplateElement from './UserFormTemplateElement.vue';

import { wrapperFactory } from '@/common/test';
import { USER_TEMPLATES } from '@/user/fixtures';

const template = dataTest('user-form-template-element');

let wrapper: VueWrapper<InstanceType<typeof UserFormTemplateList>>;

beforeEach(() => {
  wrapper = wrapperFactory(UserFormTemplateList, { templates: USER_TEMPLATES });
});

enableAutoUnmount(afterEach);

describe('UserFormTemplateList', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(UserFormTemplateList)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows templates', async () => {
    expect(wrapper.findAll(template).length).toBe(USER_TEMPLATES.length);
    expect(wrapper.findComponent<typeof UserFormTemplateElement>(template).props('template')).toStrictEqual(
      USER_TEMPLATES[0]
    );
  });

  it('emits edit template', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('edit');

    wrapper.findComponent<typeof UserFormTemplateElement>(template).vm.$emit('edit', USER_TEMPLATES[0]);

    expect(wrapper.emitted('edit')).toHaveLength(1);
    expect(wrapper.emitted()['edit'][0]).toStrictEqual([USER_TEMPLATES[0]]);
  });

  it('emits delete template', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('delete');

    wrapper.findComponent<typeof UserFormTemplateElement>(template).vm.$emit('delete', USER_TEMPLATES[0]._id);

    expect(wrapper.emitted('delete')).toHaveLength(1);
    expect(wrapper.emitted()['delete'][0]).toStrictEqual([USER_TEMPLATES[0]._id]);
  });
});
