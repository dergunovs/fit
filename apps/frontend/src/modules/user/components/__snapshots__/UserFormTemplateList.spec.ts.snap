// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserFormTemplateList > matches snapshot 1`] = `
"<div wrap="">
  <user-form-template-element-stub template="[object Object]" data-test="user-form-template-element"></user-form-template-element-stub>
</div>"
`;
